OpenLiberty/open-liberty
# Netty JMS

Updated the JMS Client bundle to use Netty framework. This is a draft prototype of a working implementation that can change on an endpoint by endpoint basis to use either Netty or CHFW including beta fencing and metatype parsing. There are a couple of TODOs and discussion to be done about the changes such as

**Netty Related**
- Verify endpoint access **Discussed** Fine but need to check that everything gets cleaned up appropriately. First need to check if the shutdown and clean up is called appropriately. Check using the Concurrent in Netty bundle.
  - Discuss with both teams about how to access endpoints with their ID to match to the appropriate configuration. Right now added map from id to chain but don't think that's the best way to do this
- SSL Support **Discussed**
  - Need to discuss code sharing between components and properly verify that security is established properly not sure if we can check with a security focal point to assist with this
  - Further testing
- Virtual Connection equivalents
  - Need to verify equivalent functionality like `requestPermissionToClose` on VirtualConnection to Netty which does not have a direct method call that does the same thing
- Common bundle usage **Discussed** Check common approaches with handlers and using the bundle as a framework so that we use the same pipelines between all components.
  - Setting TCP options on endpoints
  - Investigating shared common handler approaches
  - Investigate bundle logging
- OSGi bundle loading
  - Need to investigate if/how to load bundles optionally according to metatype/beta functionality to load only the bundles that should be used or something equivalent to follow convention [here](https://github.com/OpenLiberty/open-liberty/wiki/Beta-Fencing)
- Metatype beta fencing
  - ~Check about removing the ibm:beta property~ Remove it since it is internal and will be beta fenced. Only used primarily for documentation


**JMS Related**
- Different connection types
  - It seems like there are three different ways to start a connection with the JMS client as seen [here](https://github.com/OpenLiberty/open-liberty/blob/d71325f62fda5d99375abc82a77aec13819dcf48/dev/com.ibm.ws.messaging.comms.client/src/com/ibm/ws/sib/jfapchannel/impl/octracker/ConnectionDataGroup.java#L571-L658) out of which I've only seen one being used on the tests I've ran. Need to investigate if the other connection types are used and how to test the changes there.
- Connection invalidation
  - There may be cases where the connection is dropped and need to investigate what would be the best way to invalidate the connection and close it appropriately
- Heartbeat testing
  - Further test heartbeat works as expected specifically what happens if it can be configure programmatically and if the timeouts are happening as expected
  - Check if heartbeat functionality in the new handler can be included in the client handler as one
- Verify Additional testing besides OL FATs


fixes #20210 #20132
